python/dgl/data/adapter.py

```diff
@@ -60,48 +65,233 @@ class AsNodePredDataset(DGLDataset):
 
     def has_cache(self):
-        return os.path.isfile(os.path.join(self.save_path, 'graph.bin'))
+        return os.path.isfile(os.path.join(self.save_path, 'graph_{}.bin'.format(self.hash)))
 
     def load(self):
-        with open(os.path.join(self.save_path, 'info.json'), 'r') as f:
+        with open(os.path.join(self.save_path, 'info_{}.json'.format(self.hash)), 'r') as f:
             info = json.load(f)
             if (info['split_ratio'] != self.split_ratio
-                or info['target_ntype'] != self.target_ntype):
+                    or info['target_ntype'] != self.target_ntype):
                 raise ValueError('Provided split ratio is different from the cached file. '
                                  'Re-process the dataset.')
             self.split_ratio = info['split_ratio']
             self.target_ntype = info['target_ntype']
             self.num_classes = info['num_classes']
-        gs, _ = utils.load_graphs(os.path.join(self.save_path, 'graph.bin'))
+        gs, _ = utils.load_graphs(os.path.join(self.save_path, 'graph_{}.bin'.format(self.hash)))
+        self.g = gs[0]
+
+    def save(self):
+        utils.save_graphs(os.path.join(self.save_path, 'graph_{}.bin'.format(self.hash)), [self.g])
+        with open(os.path.join(self.save_path, 'info_{}.json'.format(self.hash)), 'w') as f:
+            json.dump({
+                'split_ratio': self.split_ratio,
+                'target_ntype': self.target_ntype,
+                'num_classes': self.num_classes}, f)
+
+    def __getitem__(self, idx):
+        return self.g
+
+    def __len__(self):
+        return 1
+
+
+def negative_sample(g, num_samples):
+    """Random sample negative edges from graph, excluding self-loops"""
+    num_nodes = g.num_nodes()
+    redundancy = _calc_redundancy(
+        num_samples, g.num_edges(), num_nodes ** 2)
+    sample_size = int(num_samples*(1+redundancy))
+    edges = np.random.randint(0, num_nodes, size=(2, sample_size))
+    edges = np.unique(edges, axis=1)
+    # remove self loop
+    mask_self_loop = edges[0] == edges[1]
+    # remove existing edges
+    has_edges = F.asnumpy(g.has_edges_between(edges[0], edges[1]))
+    mask = ~(np.logical_or(mask_self_loop, has_edges))
+    edges = edges[:, mask]
+    if edges.shape[1] >= num_samples:
+        edges = edges[:, :num_samples]
+    return edges
+
+
+class AsEdgePredDataset(DGLDataset):
+    """Repurpose a dataset for edge prediction task.
+
+    The created dataset will include data needed for link prediction.
+    Currently only support homogeneous graph. 
+    It will keep only the first graph in the provided dataset and
+    generate train/val/test edges according to the given split ratio,
+    and the correspondent negative edges based on the neg_ratio. The generated
+    edges will be cached to disk for fast re-loading. If the provided split ratio
+    differs from the cached one, it will re-process the dataset properly.
+
```

Comment:
Only support dataset of homogeneous graphs.

Fix numpy random seed to generate deterministic train/val/test split.